kernel/term_typing.ml

```diff
@@ -62,69 +63,106 @@ let feedback_completion_typecheck =
 
 type typing_context =
 | MonoTyCtx of Environ.env * unsafe_type_judgment * Id.Set.t * Stateid.t option
-| PolyTyCtx of Environ.env * unsafe_type_judgment * Univ.universe_level_subst * Univ.AUContext.t * Id.Set.t * Stateid.t option
+| PolyTyCtx of Environ.env * unsafe_type_judgment * universe_level_subst * AUContext.t * Id.Set.t * Stateid.t option
 
-let infer_declaration env (dcl : constant_entry) =
-  match dcl with
-    | ParameterEntry (ctx,(t,uctx),nl) ->
-      let env = match uctx with
-        | Monomorphic_entry uctx -> push_context_set ~strict:true uctx env
-        | Polymorphic_entry (_, uctx) -> push_context ~strict:false uctx env
+let check_primitive_type env op_t u t =
+  let inft = Typeops.type_of_prim_or_type env u op_t in
+  try Reduction.default_conv ~l2r:false Reduction.CONV env inft t
+  with Reduction.NotConvertible ->
+    Type_errors.error_incorrect_primitive env (make_judge op_t inft) t
+
+let merge_unames =
+  Array.map2 (fun base user -> match user with Anonymous -> base | Name _ -> user)
+
+let infer_primitive env { prim_entry_type = utyp; prim_entry_content = p; } =
+  let open CPrimitives in
+  let auctx = CPrimitives.op_or_type_univs p in
+  let univs, typ =
+    match utyp with
+    | None ->
+      let u = UContext.instance (AUContext.repr auctx) in
+      let typ = Typeops.type_of_prim_or_type env u p in
+      let univs = if AUContext.is_empty auctx then Monomorphic ContextSet.empty
+        else Polymorphic auctx
       in
-      let j = Typeops.infer env t in
-      let usubst, univs = Declareops.abstract_universes uctx in
-      let r = Typeops.assumption_of_judgment env j in
-      let t = Vars.subst_univs_level_constr usubst j.uj_val in
-      {
-        Cooking.cook_body = Undef nl;
-        cook_type = t;
-        cook_universes = univs;
-        cook_relevance = r;
-        cook_inline = false;
-        cook_context = ctx;
-      }
+      univs, typ
 
-    (** Primitives cannot be universe polymorphic *)
-    | PrimitiveEntry ({ prim_entry_type = otyp;
-                        prim_entry_univs = uctxt;
-                        prim_entry_content = op_t;
-                      }) ->
-      let env = push_context_set ~strict:true uctxt env in
-      let ty = match otyp with
-      | Some typ ->
+    (* TODO? factor monomorphic and polymorphic cases? *)
```

Comment:
@SkySkimmer do you think this should be done now?

Reply:
This already looks reasonably factored, I forgot what more I wanted.